Lesson17/main.js

```diff
@@ -0,0 +1,132 @@
+const slideList = document.getElementById("js-slid-list");
+const slideWrapper = document.getElementById("js-slide-wrapper");
+
+const createElementWithClassName = (type, className) => {
+  const element = document.createElement(type);
+  element.className = className;
+  return element;
+}
+
+const addLoading = (parent) => {
+  const loading = createElementWithClassName("img", "loading");
+  loading.id = "js-loading";
+  loading.src = "./loading-circle.gif";
+  parent.appendChild(loading);
+}
+
+const removeLoading = () => document.getElementById("js-loading").remove();
+
+const addErrorMessage = (error, parent) => {
+  const errorMessage = createElementWithClassName("p", "error-message");
+  errorMessage.textContent = error;
+  parent.appendChild(errorMessage);
+}
+
+const fetchErrorHandling = async (response) => {
+  if (response.ok) {
+    return await response.json();
+  } else {
+    throw new Error("サーバーエラーが発生しました");
+  }
+}
+
+const getJsonOrError = async (url) => {
+  const response = await fetch(url);
+  const json = await fetchErrorHandling(response);
+  return json;
+}
+
+const getSlideImageData = async () => {
+  return new Promise((resolve) => {
+    setTimeout(() => resolve(getJsonOrError("https://myjson.dit.upm.es/api/bins/gzcn")), 3000);
+  })
+}
+
+const tryGetSlideImageData = async () => {
+  addLoading(slideWrapper);
+  try {
+    return await getSlideImageData();
+  } catch (e) {
+    addErrorMessage(e, slideWrapper);
+  } finally {
+    removeLoading();
+  }
+}
+
+const init = async () => {
+  const imageData = await tryGetSlideImageData();
+  if(imageData) {
+    initSlideItem(imageData); 
+    initArrowButtons();
+  }
+}
+
+const initSlideItem = (imageSrc) => {
+  for (let i = 0; i < imageSrc.length; i++) {
+    const createdSlideItem = createSlideItem(imageSrc[i]);
+    //初期設定として 最初の要素にis-displayingクラスを付与
+    i === 0 && createdSlideItem.classList.add("is-displaying");
+    addElement(createdSlideItem, slideList);
+  }
+}
+
+const initArrowButtons = () => {
+  const createdArrowButtons = createArrowButtons();
+  addElement(createdArrowButtons, slideWrapper);
+  setClickEventInArrowButton();
+}
+
+const createSlideItem = ({ image }) => {
+  const slideItem = createElementWithClassName("li", "slide-item");
+  const slideImage = createElementWithClassName("img", "slide-img");
+  slideImage.src = image;
+  slideItem.appendChild(slideImage);
+  return slideItem;
+}
+
+const addElement = (element, parent) => parent.appendChild(element);
+
```

Comment:
I thought `addElement` you implemented this time can be used for other functions :)

Reply:
Thank you.
I should have looked a little more carefully...

Thanks to yuka, I was able to separate the functions that create elements from the functions that add them!
fec154d

Reply:
Oh sorry, I meant I thought you can use `addElement` in other functions.
For example, on `createSlideItem` function.

```
const createSlideItem = ({ image }) => {
  const slideItem = createElementWithClassName("li", "slide-item");
  const slideImage = createElementWithClassName("img", "slide-img");
  slideImage.src = image;
  slideItem.appendChild(slideImage);  ←here "addElement(slideImage,slideItem)" like this
  return slideItem;
};
```
But this is just IMO, I also thought you might have some idea and separated it depending on the part. So it's up to you :)

Reply:
Thank you very much.
I noticed that after reading yuka's review.
I didn't think addElement was necessary...
It does the same thing as appendChild, so I didn't think it needed to be a function.

I'm sorry to keep repeating this, but please confirm.
5c4db7d